patchShebangs: Allow multiple args

JohnAZoidberg opened the pull request:
###### Motivation for this change
It's tempting to think patchShebangs supports multiple arguments.
Without this patch it just silently ignores all but the first. Now it
patches the shebangs in all of its arguments.

Fixes: #57695

Obviously I haven't tested it on nixpkgs because it rebuilds *everything*. Please check whether there might be some edge cases that I forgot which could break with these changes.
I tried the function on its own (copied tothe `patchPhase` of a derivation) and it seems to work as intended.

Seems like there's no docbook documentation for this function. I think I'll do that next.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

JohnAZoidberg committed:
patchShebangs: Explain that script must be executable

JohnAZoidberg committed:
patchShebangs: Allow for multiple arguments

It's tempting to think patchShebangs supports multiple arguments.
Without this patch it just silently ignores all but the first. Now it
patches the shebangs in all of its arguments.

Fixes: #57695

matthewbauer commented on pkgs/build-support/setup-hooks/patch-shebangs.sh:
```diff
@@ -29,75 +32,76 @@ patchShebangs() {
+                  rm "$timestamp"
+              fi
+          fi
+      done < <(find "$dir" -type f -perm -0100 -print0)
```
Could you just put ```find "$@"``` here? I think that could be more efficient

JohnAZoidberg replied:
Oh, that would remove the need for my loop, right?

matthewbauer replied:
Yeah would just have to remove the ```[ -e "$dir" ] || return 0``` part as well.

JohnAZoidberg replied:
Thanks! I tried the function out on its own and it seems to be doing the right things.

FRidh commented:
Waiting for reviews before merging.

chreekat commented on pkgs/build-support/setup-hooks/patch-shebangs.sh:
```diff
@@ -95,7 +94,7 @@ patchShebangs() {
             fi
         fi
-    done < <(find "$dir" -type f -perm -0100 -print0)
+    done < <(find "$@" -type f -perm -0100 -print0)
```
This change, combined with the removal of line 43, means the function will fail clumsily if no arguments are passed to it. But considering that the old version silently and surprisingly ignores nonexistent files, this could be considered an improvement. ;) However, since it's a change in behavior, might it cause build failures? Should we let them happen? (Serious question.)

JohnAZoidberg replied:
What about we check if there are no arguments supplied in line 43?
```bash
  if [ $# -eq 0 ]; then
      echo "No arguments supplied to patchShebangs" >&2
      return 3141592  # Not sure which return code is best, maybe just 1
  fi
```

JohnAZoidberg replied:
I made it exit with 0 like it was before so that existing scripts don't break.

JohnAZoidberg committed:
patchShebangs: Exit if no arguments were given

Commit "patchShebangs: Allow for multiple arguments" 4a1e51f957b57f78
removed the check. We don't want to break existing usages so this
introduces it again with a successful exit code.

JohnAZoidberg commented:
@matthewbauer Do you still approve?
Does anyone else have any thoughts?
@chreekat Does this address your concerns?

chreekat commented:
Lgtm, yep

matthewbauer commented on pkgs/build-support/setup-hooks/patch-shebangs.sh:
```diff
@@ -40,7 +41,10 @@ patchShebangs() {
-    [ -e "$dir" ] || return 0
+    if [ $# -eq 0 ]; then
+        echo "No arguments supplied to patchShebangs" >&2
+        exit 0
```
This should be >0 so bash recognizes it as an error

matthewbauer replied:
Should be return actually

Infinisil replied:
Why `>0`? That just pipes the string to a file with name "0"?

matthewbauer replied:
Oops meant "greater than 0". That is non-zero code.

JohnAZoidberg replied:
ohhhh

matthewbauer commented on pkgs/build-support/setup-hooks/patch-shebangs.sh:
```diff
@@ -29,9 +32,7 @@ patchShebangs() {
-    local dir="$1"
-
-    header "patching script interpreter paths in $dir"
+    echo "patching script interpreter paths in $@"
```
I think header should be kept too

JohnAZoidberg replied:
But it just calls echo and is marked as obsolete. https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L115

Infinisil commented:
https://github.com/NixOS/nixpkgs/pull/60559#discussion_r287837965 ?
